File: Hash-Based Searching/Open Addressing Search/Linear Probing/LinearProbing.js
class LinearProbingHashTable {

    static TABLE_SIZE = 11;
    static EMPTY = -1;

    constructor() {
        this.table = Array(
            LinearProbingHashTable.TABLE_SIZE
        ).fill(
            LinearProbingHashTable.EMPTY
        );
    }

    hashFunction(key) {
        return (
            key %
            LinearProbingHashTable.TABLE_SIZE
        );
    }

    insert(key) {

        const index =
            this.hashFunction(key);

        for (
            let i = 0;
            i <
            LinearProbingHashTable.TABLE_SIZE;
            i++
        ) {

            const pos =
                (index + i) %
                LinearProbingHashTable.TABLE_SIZE;

            if (
                this.table[pos] ===
                LinearProbingHashTable.EMPTY
            ) {
                this.table[pos] = key;
                return true;
            }
        }

        return false;
    }

    search(key) {

        const index =
            this.hashFunction(key);

        for (
            let i = 0;
            i <
            LinearProbingHashTable.TABLE_SIZE;
            i++
        ) {

            const pos =
                (index + i) %
                LinearProbingHashTable.TABLE_SIZE;

            if (
                this.table[pos] ===
                LinearProbingHashTable.EMPTY
            )
                return false;

            if (
                this.table[pos] === key
            )
                return true;
        }

        return false;
    }

    display() {

        console.log("\nHash Table");

        for (
            let i = 0;
            i <
            LinearProbingHashTable.TABLE_SIZE;
            i++
        ) {

            console.log(
                `[${i}] = ${
                    this.table[i] ===
                    LinearProbingHashTable.EMPTY
                        ? "EMPTY"
                        : this.table[i]
                }`
            );
        }
    }
}

const ht =
    new LinearProbingHashTable();

ht.insert(22);
ht.insert(33);
ht.insert(44);
ht.insert(55);

ht.display();

console.log(
    "\nSearch 44:",
    ht.search(44)
        ? "Found"
        : "Not Found"
);

File: Hash-Based Searching/Open Addressing Search/Linear Probing/LinearProbing.cpp
#include <iostream>
#include <vector>

class LinearProbingHashTable {
private:
    static constexpr int TABLE_SIZE = 11;
    static constexpr int EMPTY = -1;

    std::vector<int> table;

    int hashFunction(int key) const {
        return key % TABLE_SIZE;
    }

public:
    LinearProbingHashTable()
        : table(TABLE_SIZE, EMPTY) {}

    bool insert(int key) {
        int index = hashFunction(key);

        for (int i = 0; i < TABLE_SIZE; i++) {
            int pos = (index + i) % TABLE_SIZE;

            if (table[pos] == EMPTY) {
                table[pos] = key;
                return true;
            }
        }

        return false;
    }

    bool search(int key) const {
        int index = hashFunction(key);

        for (int i = 0; i < TABLE_SIZE; i++) {
            int pos = (index + i) % TABLE_SIZE;

            if (table[pos] == EMPTY)
                return false;

            if (table[pos] == key)
                return true;
        }

        return false;
    }

    void display() const {
        std::cout << "\nHash Table\n";

        for (int i = 0; i < TABLE_SIZE; i++) {
            std::cout << "[" << i << "] = ";

            if (table[i] == EMPTY)
                std::cout << "EMPTY";
            else
                std::cout << table[i];

            std::cout << '\n';
        }
    }
};

int main() {
    LinearProbingHashTable ht;

    ht.insert(22);
    ht.insert(33);
    ht.insert(44);
    ht.insert(55);

    ht.display();

    std::cout
        << "\nSearch 44: "
        << (ht.search(44)
                ? "Found"
                : "Not Found")
        << '\n';

    return 0;
}
